wcfsetup/install/files/lib/system/box/AbstractCategoriesBoxController.class.php

```diff
@@ -0,0 +1,133 @@
+<?php
+
+namespace wcf\system\box;
+
+use wcf\data\box\Box;
+use wcf\data\category\AbstractDecoratedCategory;
+use wcf\data\category\CategoryNodeTree;
+use wcf\system\WCF;
+
+/**
+ * Abstract implmentation for category box.
+ *
+ * @author  Marcel Werk
+ * @copyright   2001-2023 WoltLab GmbH
+ * @license GNU Lesser General Public License <http://opensource.org/licenses/lgpl-license.php>
+ * @since   6.0
+ */
+abstract class AbstractCategoriesBoxController extends AbstractBoxController implements IConditionBoxController
+{
+    /**
+     * @inheritDoc
+     */
+    protected static $supportedPositions = [
+        'footerBoxes',
+        'sidebarLeft',
+        'sidebarRight',
+        'contentTop',
+        'contentBottom',
+        'footer',
+    ];
+
+    protected bool $showChildCategories = false;
+
+    /**
+     * @inheritDoc
+     */
+    protected function loadContent()
+    {
+        $categoryTree = $this->getNodeTree();
+        $categoryList = $categoryTree->getIterator();
+
+        if (\iterator_count($categoryList)) {
+            $this->content = WCF::getTPL()->fetch(
+                'boxCategories',
+                'wcf',
+                [
+                    'categoryList' => $categoryList,
+                    'activeCategory' => $this->getActiveCategory(),
+                    'resetLink' => $this->getResetLink(),
+                    'showChildCategories' => $this->showChildCategories,
+                ],
+                true
+            );
+        }
+    }
+
+    protected abstract function getNodeTree(): CategoryNodeTree;
+
+    protected function getActiveCategory(): ?AbstractDecoratedCategory
+    {
+        return null;
+    }
+
+    protected function getResetLink(): string
+    {
+        return '';
+    }
+
+    /**
+     * @inheritDoc
+     */
+    public function readConditions()
+    {
+        if (!empty($_POST['showChildCategories'])) {
+            $this->showChildCategories = true;
+        }
+    }
+
+    /**
+     * @inheritDoc
+     */
+    public function validateConditions()
+    {
+    }
+
+    /**
+     * @inheritDoc
+     */
+    public function getConditionDefinition()
```

Comment:
Please also add return types to inherited methods (where reasonably possible): Adding return types in child classes is safe from a Liskov substitution principle (“LSP”) perspective, because return types are covariant.